x/concentrated-liquidity/internal/swapstrategy/zero_for_one.go

```diff
@@ -78,17 +79,121 @@ func (s zeroForOneStrategy) ComputeSwapStep(sqrtPriceCurrent, sqrtPriceNextTick,
+	// amountOne is amount in.
+	// TODO: multiplication with rounding up at precision end.
+	feeChargeTotal := amountOne.Mul(s.swapFee).Quo(sdk.OneDec().Sub(s.swapFee))
+
+	return sqrtPriceNext, amountZero, amountOne, feeChargeTotal
+}
+
+// ComputeSwapStepOutGivenIn calculates the next sqrt price, the amount of token in consumed, the amount out to return to the user, and total fee charge on token in.
+// Parameters:
+//   - sqrtPriceCurrent is the current sqrt price.
+//   - sqrtPriceTarget is the target sqrt price computed with GetSqrtTargetPrice(). It must be one of:
+//     1. Next tick sqrt price.
+//     2. Sqrt price limit representing price impact protection.
+//   - liquidity is the amount of liquidity between the sqrt price current and sqrt price target.
+//   - amountZeroInRemaining is the amount of token zero in remaining to be swapped. This amount is fully consumed
+//     if sqrt price target is not reached. In that case, the returned amountZeroIn is the amount remaining given.
+//     Otherwise, the returned amountIn will be smaller than amountZeroInRemaining given.
+//
+// Returns:
+//   - sqrtPriceNext is the next sqrt price. It equals sqrt price target if target is reached. Otherwise, it is in-between sqrt price current and target.
+//   - amountZeroIn is the amount of token zero in consumed. It equals amountZeroInRemaining if target is reached. Otherwise, it is less than amountZeroInRemaining.
+//   - amountOutComputed is the amount of token out computed. It is the amount of token out to return to the user.
+//   - feeChargeTotal is the total fee charge. The fee is charged on the amount of token in.
+//
+// ZeroForOne details:
+// - zeroForOneStrategy assumes moving to the left of the current square root price.
+func (s zeroForOneStrategy) ComputeSwapStepOutGivenIn(sqrtPriceCurrent, sqrtPriceTarget, liquidity, amountZeroInRemaining sdk.Dec) (sdk.Dec, sdk.Dec, sdk.Dec, sdk.Dec) {
+	// Estimate the amount of token zero needed until the target sqrt price is reached.
+	amountZeroIn := math.CalcAmount0Delta(liquidity, sqrtPriceTarget, sqrtPriceCurrent, true) // N.B.: if this is false, causes infinite loop
+
+	// Calculate sqrtPriceNext on the amount of token remaining after fee.
+	amountZeroInRemainingLessFee := amountZeroInRemaining.Mul(sdk.OneDec().Sub(s.swapFee))
+	var sqrtPriceNext sdk.Dec
+	// If have more of the amount remaining after fee than estimated until target,
+	// bound the next sqrtPriceNext by the target sqrt price.
+	if amountZeroInRemainingLessFee.GTE(amountZeroIn) {
+		sqrtPriceNext = sqrtPriceTarget
 	} else {
-		// amountOne is amount in.
-		// TODO: multiplication with rounding up at precision end.
-		feeChargeTotal = amountOne.Mul(s.swapFee).Quo(sdk.OneDec().Sub(s.swapFee))
+		// Otherwise, compute the next sqrt price based on the amount remaining after fee.
+		sqrtPriceNext = math.GetNextSqrtPriceFromAmount0InRoundingUp(sqrtPriceCurrent, liquidity, amountZeroInRemainingLessFee)
```

Comment:
Why do we round up here but round down in the equivalent function in `one_for_zero.go`?

Reply:
I have this documented in the spec for the method:

When we swap for token one out given token zero in, the price is decreasing, and we need to move the sqrt price (decrease it) less o avoid overpaying the amount out of the pool. Therefore, we round up.

Similar documentation exists for the round down equivalent if you are interested